

diff --git a/memory/volatile/VolatileBufferWindows.cpp b/memory/volatile/VolatileBufferWindows.cpp
index fbd4eec07a0e3..6615f89299e6e 100644
--- a/memory/volatile/VolatileBufferWindows.cpp
+++ b/memory/volatile/VolatileBufferWindows.cpp
@@ -6,6 +6,7 @@
 #include "mozilla/Assertions.h"
 #include "mozilla/DebugOnly.h"
 #include "mozilla/mozalloc.h"
+#include "mozilla/WindowsVersion.h"
 
 #include <windows.h>
 
@@ -39,6 +40,11 @@ bool VolatileBuffer::Init(size_t aSize, size_t aAlignment) {
     goto heap_alloc;
   }
 
+  static bool sUndoSupported = IsWin8OrLater();
+  if (!sUndoSupported) {
+    goto heap_alloc;
+  }
+
   mBuf = VirtualAllocEx(GetCurrentProcess(), nullptr, mSize,
                         MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
   if (mBuf) {
