

diff --git a/mozglue/misc/PreXULSkeletonUI.cpp b/mozglue/misc/PreXULSkeletonUI.cpp
index 18fe16c0d8f95..a524477d12da7 100644
--- a/mozglue/misc/PreXULSkeletonUI.cpp
+++ b/mozglue/misc/PreXULSkeletonUI.cpp
@@ -34,6 +34,7 @@
 #include "mozilla/UniquePtrExtensions.h"
 #include "mozilla/Unused.h"
 #include "mozilla/WindowsDpiAwareness.h"
+#include "mozilla/WindowsVersion.h"
 #include "mozilla/WindowsProcessMitigations.h"
 
 namespace mozilla {
@@ -1780,6 +1781,10 @@ static Result<Ok, PreXULSkeletonUIError> CreateAndStorePreXULSkeletonUIImpl(
 
   const TimeStamp skeletonStart = TimeStamp::Now();
 
+  if (!IsWin10OrLater()) {
+    return Err(PreXULSkeletonUIError::Ineligible);
+  }
+
   HKEY regKey;
   MOZ_TRY_VAR(regKey, OpenPreXULSkeletonUIRegKey());
   AutoCloseRegKey closeKey(regKey);
diff --git a/mozglue/misc/PreXULSkeletonUI.h b/mozglue/misc/PreXULSkeletonUI.h
index 7bd5b86359fc9..e7d19aea3b9f8 100644
--- a/mozglue/misc/PreXULSkeletonUI.h
+++ b/mozglue/misc/PreXULSkeletonUI.h
@@ -77,6 +77,7 @@ struct ThemeColors {
 
 enum class PreXULSkeletonUIError : uint32_t {
   None,
+  Ineligible,
   Disabled,
   EnabledKeyDoesNotExist,
   OOM,
diff --git a/mozglue/misc/WindowsDpiInitialization.cpp b/mozglue/misc/WindowsDpiInitialization.cpp
index 972d577651c34..2e957b583012b 100644
--- a/mozglue/misc/WindowsDpiInitialization.cpp
+++ b/mozglue/misc/WindowsDpiInitialization.cpp
@@ -49,7 +49,7 @@ WindowsDpiInitializationResult WindowsDpiInitialization() {
     }
 
     return WindowsDpiInitializationResult::Success;
-  } else {
+  } else if (IsWin8Point1OrLater()) {
     DynamicallyLinkedFunctionPtr<SetProcessDpiAwarenessType>
         setProcessDpiAwareness(L"Shcore.dll", "SetProcessDpiAwareness");
     if (!setProcessDpiAwareness) {
@@ -60,6 +60,12 @@ WindowsDpiInitializationResult WindowsDpiInitialization() {
       return WindowsDpiInitializationResult::SetProcessDpiAwarenessFailed;
     }
 
+    return WindowsDpiInitializationResult::Success;
+  } else {
+    if (!SetProcessDPIAware()) {
+      return WindowsDpiInitializationResult::SetProcessDPIAwareFailed;
+    }
+
     return WindowsDpiInitializationResult::Success;
   }
 }
diff --git a/mozglue/misc/WindowsDpiInitialization.h b/mozglue/misc/WindowsDpiInitialization.h
index 7379bd53e104a..5943d50526e30 100644
--- a/mozglue/misc/WindowsDpiInitialization.h
+++ b/mozglue/misc/WindowsDpiInitialization.h
@@ -16,6 +16,7 @@ enum class WindowsDpiInitializationResult : uint32_t {
   SetProcessDpiAwarenessContextFailed,
   FindSetProcessDpiAwarenessFailed,
   SetProcessDpiAwarenessFailed,
+  SetProcessDPIAwareFailed,
 };
 
 // Get a string representation of any WindowsDpiInitializationResult value
@@ -33,6 +34,8 @@ inline const char* WindowsDpiInitializationResultString(
       return "Failed to find SetProcessDpiAwareness";
     case WindowsDpiInitializationResult::SetProcessDpiAwarenessFailed:
       return "SetProcessDpiAwareness failed";
+    case WindowsDpiInitializationResult::SetProcessDPIAwareFailed:
+      return "SetProcessDPIAware failed";
     default:
       return "Unknown result";
   }
